app/src/main/java/pm/gnosis/heimdall/ui/safe/details/transactions/SafeTransactionsViewModel.kt

```diff
@@ -31,44 +29,63 @@ class SafeTransactionsViewModel @Inject constructor(
 ) : SafeTransactionsContract() {
 
     private var address: Solidity.Address? = null
+    private var cachedData: Adapter.Data<AdapterEntry>? = null
 
     override fun setup(address: Solidity.Address) {
         this.address = address
     }
 
-    override fun observeTransactions(): Flowable<out Result<Adapter.Data<String>>> {
+    override fun observeTransactions(): Flowable<out Result<Adapter.Data<AdapterEntry>>> {
         return this.address?.let {
-            safeRepository.observeTransactionDescriptions(it)
-                .scanToAdapterData()
+            Flowable.combineLatest<List<TransactionStatus>, List<TransactionStatus>, List<AdapterEntry>>(
+                safeRepository.observePendingTransactions(it),
+                safeRepository.observeSubmittedTransactions(it),
+                BiFunction { pending, submitted ->
+                    val currentTime = System.currentTimeMillis()
+                    val combined = ArrayList<AdapterEntry>((pending.size + submitted.size) * 2)
+                    if (pending.isNotEmpty()) {
+                        combined += AdapterEntry.Header(R.string.header_pending)
+                        pending.forEach { combined += AdapterEntry.Transaction(it.id) }
+                    }
+                    var lastHeaderRes = 0
+                    submitted.forEach {
+                        val timeDiff = currentTime - it.timestamp
+                        val headerRes = when {
+                            timeDiff < DateTimeUtils.DAY_IN_MS -> R.string.header_today
+                            timeDiff < 2 * DateTimeUtils.DAY_IN_MS -> R.string.header_yesterday
+                            lastHeaderRes == 0 -> R.string.header_submitted // First header should never be "older"
+                            else -> R.string.header_older
+                        }
+                        if (lastHeaderRes != headerRes && headerRes != 0) {
+                            combined += AdapterEntry.Header(headerRes)
+                            lastHeaderRes = headerRes
+                        }
+                        combined += AdapterEntry.Transaction(it.id)
+                    }
+                    combined
+                }
+            )
+                .scanToAdapterData(initialData = cachedData)
+                .doOnNext { cachedData = it }
                 .mapToResult()
-        } ?: Flowable.empty<Result<Adapter.Data<String>>>()
+        } ?: Flowable.empty<Result<Adapter.Data<AdapterEntry>>>()
```

Comment:
Scheduler also missing here (when added to the `safeRepository.observePendingTransactions` or to `safeRepository.observeSubmittedTransactions` it should run on that scheduler though...)

Reply:
added in the Repository